### PR TITLE
Simplify Time Management

### PR DESCRIPTION
**Pull request by protonspring**
This is a functional simplification that dramatically simplifies time management.  After significant mathematical modeling, these simple log equations closely follow what master was doing.  I need help and suggestions.  I don’t think this is commit-able as is, but it probably can be.  Please help me clean this up and make recommendations.

GOOD
1. Instead of 2500 calls to move_importance per move, we have a single log equation.
2.  Much less code.
3.  May result in some elo.
4.  Can more easily be tuned.
5.  Can customize certain special time scenarios like when time increment is lower than move overhead.
6.  Allocates less time for the first 10 ply and allows move time for middlegame moves.  I suspect this is where some of the performance comes from.
7. slowMover is adjusted to 100, so that it by default has no impact.  Can this be removed?

BAD
1.  I don’t like the special cases for different game types, but this may actually be a benefit because we can optimize each type of timed game.
2.  minThinkingTime breaks stuff.  If you change this value, is has a big impact.
-----It might be a good idea to change the 2*minThinkingTime to 40 or similar.
3.  I have no idea how this performs for 40/10 games.

Test results for this exact version:

STC
LLR: 2.94 (-2.94,2.94) {-1.50,0.50}
Total: 45456 W: 8647 L: 8524 D: 28285
Ptnml(0-2): 781, 5230, 10547, 5425, 745 
https://tests.stockfishchess.org/tests/view/5eae21756ffeed51f6e328a0

STC – sudden death
LLR: 2.93 (-2.94,2.94) {-1.50,0.50}
Total: 16640 W: 3944 L: 3736 D: 8960
Ptnml(0-2): 430, 1934, 3413, 2084, 459 
https://tests.stockfishchess.org/tests/view/5eae2f5f6ffeed51f6e328ad

STC – TCEC style
LLR: 2.96 (-2.94,2.94) {-1.50,0.50}
Total: 60784 W: 13459 L: 13331 D: 33994
Ptnml(0-2): 1292, 7447, 12829, 7489, 1335 
https://tests.stockfishchess.org/tests/view/5eae37346ffeed51f6e328e2

LTC
LLR: 2.94 (-2.94,2.94) {-1.50,0.50}
Total: 40208 W: 5179 L: 5112 D: 29917
Ptnml(0-2): 281, 3741, 12010, 3774, 298
https://tests.stockfishchess.org/tests/view/5eae60c76ffeed51f6e32902

Other good test results:
60+0.06: https://tests.stockfishchess.org/tests/view/5eabaee909d25e8e50581794
https://tests.stockfishchess.org/tests/view/5eab39ef09d25e8e50581756
https://tests.stockfishchess.org/tests/view/5eab24e409d25e8e50581729
https://tests.stockfishchess.org/tests/view/5ea9d73169c5cb4e2aeb8313
https://tests.stockfishchess.org/tests/view/5ea9d17d69c5cb4e2aeb830b
https://tests.stockfishchess.org/tests/view/5ea9d70e69c5cb4e2aeb8311

### CODE REVIEW

File: src/timeman.cpp [hunk 1]
```diff
@@ -105,27 +61,37 @@ void TimeManagement::init(Search::LimitsType& limits, Color us, int ply) {
...
-  {
-      // Calculate thinking time for hypothetical "moves to go"-value
-      hypMyTime =  limits.time[us]
-                 + limits.inc[us] * (hypMTG - 1)
-                 - moveOverhead * (2 + std::min(hypMTG, 40));
+  TimePoint timeLeft =  std::max(TimePoint(0),
+      limits.time[us] + limits.inc[us] * (mtg - 1) - moveOverhead * (2 + mtg));
 
-      hypMyTime = std::max(hypMyTime, TimePoint(0));
+  timeLeft = slowMover * timeLeft / 100;
```
Comment (hazzl) [inline src/timeman.cpp:71]: It might make sense to calculate std::log2(ply + 1) beforehand, since we use it twice in each branch.

**Comment by vondele** [comment]
@protonspring I'll look more carefully later, but after a recent fix a TC of 40/10 should be possible again on fishtest. Can you give it a try?

File: src/timeman.cpp (hunk 1 above)
Comment (Chess13234) [reply src/timeman.cpp:71]: An array of precalculated log2 values might be faster.

Comment (vondele) [reply src/timeman.cpp:71]: that's not really needed, this is called once per move computed. The overhead is almost 0.

**Comment by protonspring** [comment]
I will try some 40/10 tests, but i suspect they should have their own branch since the calculations are so different.

**Comment by vondele** [comment]
@protonspring more branch is really a significant disadvantage (not an advantage). I'm probably OK with the 1 branch solution in the current patch, but only to see if we can remove it afterwards. The master algorithm manages without branching, and the new solution only has to be as good as what we have right now.

**Comment by protonspring** [comment]
i may be wrong, but i think there may be some elo if we specialize time
management for each type of game.

**Comment by vondele** [comment]
@protonspring probably, but since there are too many options for the TC type (incl. base / incr) to always test exhaustively, robustness will come from something that is generic enough to deal with all those tested in a homogeneous way. I think we should optimize it for our principle choice (10+0.1 or 60+0.6) without regression otherwise. At least that's my 2cents.

**Comment by protonspring** [comment]
I agree. i will do some tests to make good performance on 40/10 then try to
generalize or consolidate.

**Comment by joergoster** [comment]
40/10"+0 doesn't make much sense, does it? You give time to play 40 moves in 10 seconds, but no time at all for the rest of the game???
This is called tournament time-control, btw, and either you give another 40/10" plus 5" or so for the rest, or you give 40/10" + 15" or something that makes sense.

**Comment by protonspring** [comment]
i had thought 40/10 was 40 moves in 10 seconds which repeats. So, 10
seconds for every 40 moves.

**Comment by joergoster** [comment]
I don't know how it is implemented in fishtest, but the running test says 40/10+0, which is bullshit, of course!

**Comment by protonspring** [comment]
i assume fishtest uses cutechess time controls. is that not correct?

**Comment by joergoster** [comment]
I see. Not very well explained in cutechess, I would say.

Setting tournament tc, or overtime https://en.wikipedia.org/wiki/Time_control, you must specify a time for the rest of the game! This can either be of sudden death, or with increment.

You can also set several "x number of moves in y time" consecutively, but you must ALWAYS set a time for the rest of the game!

**Comment by vondele** [comment]
```
  tc=TIMECONTROL	Set the time control to TIMECONTROL. The format is
			moves/time+increment, where 'moves' is the number of
			moves per tc, 'time' is time per tc (either seconds or
			minutes:seconds), and 'increment' is time increment
			per move in seconds.
```
used previously for testing as well
https://tests.stockfishchess.org/tests/view/59e1322d0ebc5916ff64c095

**Comment by MichaelB7** [comment]
Just a general observation that time control changes in the past have been a hornet's nest.  There is no adequate way for us to test for all scenarios efficiently.   I would definitely hit the pause button prior to committing.   Potential elo gains are minimal once you have something solid,  but risk is high for unintended consequences with any changes. For the most part, changes to time should be discouraged because of this.

**Comment by joergoster** [comment]
OK, so maybe I'm wrong!
Doesn't make much sense to me, though. Simply repeating x-moves/y-time endlessly is nonsense, imho.
At one point you want a game to be finished!

**Comment by protonspring** [comment]
I agree with you, but that is also a good argument that this stuff should
be simplified.

>

**Comment by protonspring** [comment]
These are well established time controls and there are other methods to
make sure games don't last forever.

**Comment by joergoster** [comment]
I agree that this kind of tc is not often used nowadays, as it has been replaced mostly by tc with increments. But anyways, we still want to support this, don't we?

**Comment by protonspring** [comment]
It doesn't seem like this us used anymore to me either. however, it is
still valid and we should still support it.

**Comment by xoto10** [comment]
IMHO the key problem here is the current code covers all kinds of time control, so any changes to it require testing all of them, when many of us only really have experience of fishtest style stc and ltc.

I can see that one "theory of everything" would be nice, but maybe there isn't one, or if there is, it's more complicated than a few separate, simple calculations? 

In my view this is an opportunity to use separate calculations for different styles of time control. This will make the code more maintainable and allow more room for future improvement.

**Comment by protonspring** [comment]
I agree. Also adding additional time controls will be much less
troublesome.

**Comment by protonspring** [comment]
Test for added handler:  x moves in y seconds (+z increment).

LLR: 2.94 (-2.94,2.94) {-1.50,0.50}
Total: 48296 W: 9553 L: 9429 D: 29314
Ptnml(0-2): 865, 5633, 11024, 5765, 861
https://tests.stockfishchess.org/tests/view/5eb1e9f62326444a3b6d3413

Should I test any other time controls?

**Comment by protonspring** [comment]
The extra code that works well for super fast games like 10+0.01 or less is partly required because of the minThinkingTime = 30;  Is there a reason for this value?  If I can lower it, the general algorithms may work better for these times.

**Comment by joergoster** [comment]
Just to mention: let's don't forget games with 'Ponder on' ... which I'm not sure if it can be tested in fishtest. And does this new code have any influence when playing with 'nodestime' like we often do in our SPSA runs?

**Comment by xoto10** [comment]
Min Thinking Time default seems to be 20 not 30? https://github.com/official-stockfish/Stockfish/blob/c527c3ad44f7465c79cef93f1e8cfebd998dc627/src/ucioption.cpp#L72

But anyway, I don't really understand having MinThinkingTime default to anything above zero. The time algo should do its thing and decide a time to use, presumably making a fairly good choice from our (devs) point of view. Then the user has the option to set MinThinkingTime if they feel that sf is moving too quickly in their setup for some reason. i.e _increase_ the value.
Clearly if it is > increment then the game will end quicker, but that is what they asked for. On the other hand, if the user hasn't specified a value I don't think MinThinkingTime should be possibly larger than increment, or affecting sudden death games in a bad way.

I suggest you change the default to zero.

**Comment by xoto10** [comment]
Hmmm, of course you'd have to check that was ok :) ...

MinThinkingTime of 20 is interesting to me since I've recently been trying really vvstc test games of 0.5+0.02 so that makes the increment the same as MinThinkingTime which seemed a bad idea. But:

I just tried dropping the default to 0 in the current code and it's awful at 0.5+0.02 vs master. (+117 -523 =160) The code in search.cpp will regularly cut off the search at not much over optimum/4 and at this vvstc I guess it's better to play at roughly 0.02s per move than to only use 0.005s on a regular basis.
So I withdraw any claim to knowledge in this area, "I know naaathing" :)

**Comment by protonspring** [comment]
possibly minThinking time is set to 30 because it improves performance at
very fast games.

This should be handled by the time management algorithms instead leaving
minThinkingTime as a user adjustment.  If a user sets minThinkngTime they
should understand that it may affect performance of very fast games.

**Comment by protonspring** [comment]
I am hoping that simplifying this time code will make all of this more
understandable and accessible to more devs.

**Comment by vondele** [comment]
> Just to mention: let's don't forget games with 'Ponder on' ... which I'm not sure if it can be tested in fishtest. 

It leaves the ponder code unchanged so I'm fine with it. Unrelated to this patch, but ponder really is rather primitive in its implementation. This is partially because pondering changes can't be tested on fishtest. There is an open issue https://github.com/glinscott/fishtest/issues/503

> And does this new code have any influence when playing with 'nodestime' like we often do in our SPSA runs?

Unlikely that nodestime is influenced much, it is also not so important as it is only (sometimes) used for SPSA tuning.

Let's not raise the bar for the patch unnecessarily, first try to get the core of it right.

**Comment by protonspring** [comment]
BOOM.  Looks like I was right.  If I reduce the influence on minThinkingTime and moveOverhead the default algorithms work VERY well for super short games.

LLR: 2.95 (-2.94,2.94) {-1.50,0.50}
Total: 6048 W: 1532 L: 1301 D: 3215
Ptnml(0-2): 141, 676, 1217, 791, 199 
https://tests.stockfishchess.org/tests/view/5eb2ccd12326444a3b6d34d6

LLR: 2.95 (-2.94,2.94) {-1.50,0.50}
Total: 3344 W: 824 L: 617 D: 1903
Ptnml(0-2): 46, 357, 710, 462, 97 
https://tests.stockfishchess.org/tests/view/5eb2d3872326444a3b6d34da

If these values are OK for minThinkingTime = 0, and moveOverhead = 10, I will commit these changes to this branch.  Can someone confirm this?

**Comment by vondele** [comment]
I would definitely be fine with default minThinkingTime 0. 

I would be more careful with moveOverhead, at least I would not change it that drastically since it is a reasonable value for normal TC. What about leaving it at 30ms default, but use moveOverhead=std::max(10, std::min(incr / 2, moveOverhead))
basically, adjust moveOverhead if it is unreasonable compared to TC.

**Comment by protonspring** [comment]
OK.  I'll run some tests.

**Comment by xoto10** [comment]
Yes, my 0.5+0.02 tests worked great when I reduced MoveOverhead to 5 or 10 and minThinkingTime to 0. (This was with master code) I agree with Vondele's suggestion for moveOverhead.

**Comment by protonspring** [comment]
Dynamic overhead seems to work just as well. . . this only change moveOverhead if the increment is very low.

https://tests.stockfishchess.org/tests/view/5eb318c32326444a3b6d350d

**Comment by vondele** [comment]
nice... 

so, now I would hope we can get rid of the special case for limits.movestogo, at least to the extend possible. Ideally we only have scale_min and scale_max set approriately for limits.movestogo zero and non-zero, and share the min/max expressions. like
```c++
     if (limits.movestogo == 0)
     {
      scale_min = std::max(2.0, 8.2 * (9.2 - std::log2(ply + 1)));
      scale_max = std::max(0.5, 1.7 * (8.0 - std::log2(ply + 1)));
     } else {
      scale_min = ...
      scale_max = ...
    }
      optimumTime = std::min<int>(0.2 * limits.time[us], timeLeft / scale_min);
      optimumTime = std::max<int>(minThinkingTime, optimumTime);

      maximumTime = std::min<int>(0.8 * limits.time[us] - moveOverhead, timeLeft / scale_max);
      maximumTime = std::max<int>(minThinkingTime, maximumTime);
```

**Comment by protonspring** [comment]
OK.  I will work on that, but there is still much to do. correct?

I haven't tested any VLTC, and there is still some regression for 40/10 games with large increments (and probably others).

**Comment by vondele** [comment]
Let's see if it is possible. Potentially N/X+Y is really different X+Y. Especially as N/X+Y reaches time control and N is very small (1,2,3 ..).

I've probably lost a bit count of the tests, so once the final version of the patch crystallizes, we need to see what we have so far. I would expect tests on 10+0.1, 60+0.6, 10+0, 10+0.01, 40/10. Maybe we want 60+0 (the most popular TC at lichess....) ?

**Comment by protonspring** [comment]
OK. I will focus on those TC's and I think I can do it, but there are so many others where there is some slight regression. .maybe.

**Comment by vondele** [comment]
Actually one more potentially needed case would be our standard threaded case 5+0.05 @ 8t 

However, eventually we need to limit the number of cases tested, accept it as a cleanup, and move to experiments aiming at Elo gain by using some characteristics of the position rather than simple ply counts as we have always had.

**Comment by MichaelB7** [comment]
Appreciate you folks taking the extra time to do additional testing.

**Comment by protonspring** [comment]
Does fish test also support moves/time+increment where time per TC could also be "minutes:seconds?" This is supported by cutechess.

Like: 60/2:30+1.5  (60 moves for every 2 minutes and 30 seconds with 1.5 seconds per move).

**Comment by vondele** [comment]
will probably break, but you could run 50/150+1.5

**Comment by protonspring** [comment]
Test supporting updated values:

LLR: 2.93 (-2.94,2.94) {-1.50,0.50}
Total: 13616 W: 2697 L: 2531 D: 8388
Ptnml(0-2): 201, 1545, 3200, 1611, 251 
https://tests.stockfishchess.org/tests/view/5eb570742326444a3b6d3dc6

**Comment by protonspring** [comment]
Linear version of x moves in y seconds (+ increment).

LLR: 2.94 (-2.94,2.94) {-1.50,0.50}
Total: 20520 W: 4099 L: 3938 D: 12483
Ptnml(0-2): 342, 2357, 4728, 2464, 369 
https://tests.stockfishchess.org/tests/view/5eb970b27dd5693aad4e6085

**Comment by protonspring** [comment]
Update.  The current version is formatted as requested and passes everything quite well except for LTC. . . still working.

**Comment by protonspring** [comment]
My last version is doing well at all time controls I can test.  Should I complete it here, or make a new clean PR ?

**Comment by vondele** [comment]
Excellent. 

I think we can complete it here or in a new PR, as you prefer. What I would need are

* Ideally, the whole sequence of commits reduced to a single one 'the final PR version', and force pushed here or as a new branch.

* The full list of passed test results (normal format, including links) for all test done on that version of the code (only), basically an up-to-date commit message.

**Comment by protonspring** [comment]
Looks pretty good so far.  Last time I tried to squash like that I almost killed myself.  I'll think I'll make a new PR with a new commit message, explanations and complete results.

**Comment by protonspring** [comment]
Updated branch.  #2678

**Pull request closed without merging**